server/src/main/java/org/elasticsearch/health/HealthService.java

```diff
@@ -61,15 +71,16 @@ public HealthService(
 
     /**
      * Returns the list of HealthIndicatorResult for this cluster.
+     * @param client A client to be used to fetch the health data from the health node
      * @param indicatorName If not null, the returned results will only have this indicator
      * @param explain Whether to compute the details portion of the results
      * @return A list of all HealthIndicatorResult if indicatorName is null, or one HealthIndicatorResult if indicatorName is not null
      * @throws ResourceNotFoundException if an indicator name is given and the indicator is not found
      */
-    public List<HealthIndicatorResult> getHealth(@Nullable String indicatorName, boolean explain) {
+    public List<HealthIndicatorResult> getHealth(Client client, @Nullable String indicatorName, boolean explain) {
         // Determine if cluster is stable enough to calculate health before running other indicators
         List<HealthIndicatorResult> preflightResults = preflightHealthIndicatorServices.stream()
-            .map(service -> service.calculate(explain))
+            .map(service -> service.calculate(explain, HealthInfo.EMPTY_HEALTH_INFO))
```

Comment:
I am wondering if we should overload this method to provide a version without the  `HealthInfo.EMPTY_HEALTH_INFO`. Since here we know for sure that we are not going to have it. I only have a vague ideas about this, but let me know if you want to brainstorm together and sketch something out.

Reply:
I went back and forth so many times on this, and honestly can't remember now why I landed on just a single method. Really it would probably make sense for preflight indicators to have their own separate interface, but I didn't want to make such a big change. I'll try out two methods locally and see if I can remember why I abandoned that (or maybe there was no good reason).

Reply:
I completely understand, that would be my first approach but I wasn't sure if it would look good ;).

Reply:
I'm starting to remember why, and maybe you have some better ideas:

- If I just add a 2nd `calculate` to the interface, then I have to implement two methods everywhere with very little gain
- If I add a 2nd `calculate` to the interface and have the 1-arg version have a default implementation then I don't have to have two implementations in each class. And it cleans up HealthService slightly. But it adds confusion (can I call the disk health indicator without HealthInfo? Or do I need to implement that method to throw an exception for the disk health indicator?), and it's barely more than a convenience on HealthService.
- I could add a 2nd `calculate` to the interface and have the 2-arg version throw an exception by default so then I don't have to have two implementations in each class. But it's got the same drawbacks as the previous one.
- I could break this into two separate interfaces -- UsesHealthInfoHealthIndicatorService and DoesNotUseHealthInfoHealthIndicatorService (not real names -- maybe just PreflightHealthIndicatorService and NonPreflightHealthIndicatorService). Maybe take `calculate` out of `HealthIndicatorService` and have them both extend that. UsesHealthInfoHealthIndicatorService has a 2-arg `calculate` method and `DoesNotUseHealthInfoHealthIndicatorService` has a 1-arg calculate method. Then we'd maybe have HealthService decide what to do with each one? Something along those lines is probably my preference, but seemed more invasive than I'd intended with this PR.

Let me know if you've got some better ideas or want to chat tomorrow.

Reply:
Something like this maybe? https://github.com/masseyke/elasticsearch/pull/6/files

Reply:
Since we have the EMPTY HealthInfo let's go with one method for now (let's allow for more indicators to develop and then have a clearer way to evolve the interface)

Reply:
Does that work for you @gmarouli? Or should we talk it through?

Reply:
Yes, it works, thank you for listing your thought process, it was very helpful 👍 !